Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -43,117 +101,147 @@ public PartialDecodingStreamReader(Task<TextReader> textReaderTask, Func<string,
+            // At the moment it's scattered around the code, and that's not great.
+            // (Additionally, we have no way of disposing of the reader if the user doesn't iterate.)
+            if (cancellationToken.IsCancellationRequested)
+            {
+                MaybeDispose();
+                cancellationToken.ThrowIfCancellationRequested();
+            }
 
-        var buffer = new char[8000];
-        while (_readyResults.Count == 0 && !_arrayClosed)
-        {
-            var taskRead = _textReader.ReadAsync(buffer, 0, buffer.Length);
-            var cancellationTask = Task.Delay(-1, cancellationToken);
-            var resultTask = await Task.WhenAny(taskRead, cancellationTask).ConfigureAwait(false);
+            if (_queuedResponses.Count > 0)
+            {
+                Current = _queuedResponses.Dequeue();
+                return true;
+            }
 
-            if (resultTask == cancellationTask)
+            // Throwing an exception takes priority over marking the sequence as completed.
+            // But if there were valid responses *before* we detected an error, we'll return those first.
+            if (_errorMessage is string)
             {
-                // If the cancellationTask "wins" `Task.WhenAny` by being cancelled, the following await will throw TaskCancelledException.
-                await cancellationTask.ConfigureAwait(false);
+                MaybeDispose();
+                // TODO: Should this actually be an RpcException?
+                throw new InvalidOperationException(_errorMessage);
             }
 
-            var readLen = await taskRead.ConfigureAwait(false);
-            if (readLen == 0)
+            if (_endOfData)
             {
-                var errorText = "Closing `]` bracket not received after iterating through the stream. " +
-                                "This means that streaming ended without all objects transmitted. " +
-                                "It is likely a result of server or network error.";
-                throw new InvalidOperationException(errorText);
+                MaybeDispose();
+                return false;
             }
 
-            var readChars = buffer.Take(readLen);
-            foreach (var c in readChars)
+            // We don't currently know what to do, so read more data and see whether that
+            // satisfies one of the above conditions. (We may need to loop multiple times.)
+            await ReadAndProcessData(cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <summary>
+    /// Called when we don't have a queued response and haven't reached a terminal state (error or completed).
+    /// </summary>
+    /// <returns></returns>
+    private async Task ReadAndProcessData(CancellationToken cancellationToken)
+    {
+        try
+        {
+            _textReader ??= await AwaitWithCancellation(_textReaderTask, cancellationToken, defaultValue: null).ConfigureAwait(false);
```

Comment:
I think that if you get rid of `_textReader` and always call `AwaitWithCancellation` here you can move the cancellation check (and the dispose call) from `MoveNext` to `AwaitWithCancellation`. You can throw the cancelled exception in AwaitWithCancellation.

The result is slightly different in that we won't throw cancelled until we have to read again, but that seems OK, we usually don't check for cancellation in between syncrhonous operations.

Reply:
I don't think I'd want AwaitWithCancellation to dispose, but using a try/finally for that seems okay - and I do like the idea of not having to explicitly call ThrowIfCanceled.

One issue there is in terms of the exception we want to be thrown on cancellation. Currently we catch the exception, retain the message, and always throw InvalidOperationException. That may not be ideal, and needs more discussion, but probably in a later iteration. (I'll add a TODO.)

Reply:
Ooh, thinking about it, I'm not sure I like still returning more results (due to them being buffered) after the cancellation token has been cancelled. Will see where I get to.